File: mRemoteNG/App/Export.vb
Imports mRemoteNG.Forms
Imports mRemoteNG.App.Runtime

Namespace App
    Public Class Export
        Public Shared Sub ExportToFile(ByVal rootTreeNode As TreeNode, ByVal selectedTreeNode As TreeNode)
            Try
                Dim exportTreeNode As TreeNode
                Dim saveSecurity As New Security.Save()

                Using exportForm As New ExportForm()
                    With exportForm
                        Select Case Tree.Node.GetNodeType(selectedTreeNode)
                            Case Tree.Node.Type.Container
                                .SelectedFolder = selectedTreeNode
                            Case Tree.Node.Type.Connection
                                If Tree.Node.GetNodeType(selectedTreeNode.Parent) = Tree.Node.Type.Container Then
                                    .SelectedFolder = selectedTreeNode.Parent
                                End If
                                .SelectedConnection = selectedTreeNode
                        End Select

                        If Not exportForm.ShowDialog(frmMain) = DialogResult.OK Then Return

                        Select Case .Scope
                            Case exportForm.ExportScope.SelectedFolder
                                exportTreeNode = .SelectedFolder
                            Case exportForm.ExportScope.SelectedConnection
                                exportTreeNode = .SelectedConnection
                            Case Else
                                exportTreeNode = rootTreeNode
                        End Select

                        saveSecurity.Username = .IncludeUsername
                        saveSecurity.Password = .IncludePassword
                        saveSecurity.Domain = .IncludeDomain
                        saveSecurity.Inheritance = .IncludeInheritance
                    End With

                    SaveExportFile(exportForm.FileName, exportForm.SaveFormat, exportTreeNode, saveSecurity)
                End Using
            Catch ex As Exception
                MessageCollector.AddExceptionMessage("App.Export.ExportToFile() failed.", ex, , True)
            End Try
        End Sub

        Private Shared Sub SaveExportFile(ByVal fileName As String, ByVal saveFormat As Config.Connections.Save.Format, ByVal rootNode As TreeNode, ByVal saveSecurity As Security.Save)
            Dim previousTimerEnabled As Boolean = False

            Try
                If TimerSqlWatcher IsNot Nothing Then
                    previousTimerEnabled = TimerSqlWatcher.Enabled
                    TimerSqlWatcher.Enabled = False
                End If

                Dim connectionsSave As New Config.Connections.Save
                With connectionsSave
                    .Export = True
                    .ConnectionFileName = fileName
                    .SaveFormat = saveFormat

                    .ConnectionList = ConnectionList
                    .ContainerList = ContainerList
                    .RootTreeNode = rootNode

                    .SaveSecurity = saveSecurity
                End With

                connectionsSave.Save()
            Catch ex As Exception
                MessageCollector.AddExceptionMessage(String.Format("Export.SaveExportFile(""{0}"") failed.", fileName), ex)
            Finally
                If TimerSqlWatcher IsNot Nothing Then
                    TimerSqlWatcher.Enabled = previousTimerEnabled
                End If
            End Try
        End Sub
    End Class
End Namespace

File: mRemoteNG/App/Export.cs
using System;
using System.Linq;
using System.Runtime.Versioning;
using System.Windows.Forms;
using mRemoteNG.Config.Connections;
using mRemoteNG.Config.DataProviders;
using mRemoteNG.Config.Serializers;
using mRemoteNG.Config.Serializers.ConnectionSerializers.Csv;
using mRemoteNG.Config.Serializers.ConnectionSerializers.Json;
using mRemoteNG.Config.Serializers.ConnectionSerializers.Rdp;
using mRemoteNG.Config.Serializers.ConnectionSerializers.Xml;
using mRemoteNG.Connection;
using mRemoteNG.Container;
using mRemoteNG.Security;
using mRemoteNG.Security.Factories;
using mRemoteNG.Tree;
using mRemoteNG.Tree.Root;
using mRemoteNG.UI.Forms;


namespace mRemoteNG.App
{
    [SupportedOSPlatform("windows")]
    public static class Export
    {
        public static void ExportToFile(ConnectionInfo selectedNode, ConnectionTreeModel connectionTreeModel)
        {
            try
            {
                SaveFilter saveFilter = new();

                using (FrmExport exportForm = new())
                {
                    if (selectedNode?.GetTreeNodeType() == TreeNodeType.Container)
                        exportForm.SelectedFolder = selectedNode as ContainerInfo;
                    else if (selectedNode?.GetTreeNodeType() == TreeNodeType.Connection)
                    {
                        if (selectedNode.Parent?.GetTreeNodeType() == TreeNodeType.Container)
                            exportForm.SelectedFolder = selectedNode.Parent;
                        exportForm.SelectedConnection = selectedNode;
                    }

                    if (exportForm.ShowDialog(FrmMain.Default) != DialogResult.OK)
                        return;

                    ConnectionInfo defaultTarget = connectionTreeModel.RootNodes.First(node => node is RootNodeInfo);
                    ConnectionInfo exportTarget = exportForm.Scope switch
                    {
                        FrmExport.ExportScope.SelectedFolder => exportForm.SelectedFolder ?? defaultTarget,
                        FrmExport.ExportScope.SelectedConnection => exportForm.SelectedConnection ?? defaultTarget,
                        _ => defaultTarget
                    };

                    saveFilter.SaveUsername = exportForm.IncludeUsername;
                    saveFilter.SavePassword = exportForm.IncludePassword;
                    saveFilter.SaveDomain = exportForm.IncludeDomain;
                    saveFilter.SaveInheritance = exportForm.IncludeInheritance;
                    saveFilter.SaveCredentialId = exportForm.IncludeAssignedCredential;

                    SaveExportFile(exportForm.FileName, exportForm.SaveFormat, saveFilter, exportTarget, exportForm.IsEncrypted, exportForm.Password);
                }
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionMessage("App.Export.ExportToFile() failed.", ex);
            }
        }

        private static void SaveExportFile(string fileName,
                                           SaveFormat saveFormat,
                                           SaveFilter saveFilter,
                                           ConnectionInfo exportTarget,
                                           bool isEncrypted = false,
                                           string password = "")
        {
            try
            {
                ISerializer<ConnectionInfo, string> serializer;
                switch (saveFormat)
                {
                    case SaveFormat.mRXML:
                        if (isEncrypted)
                        {
                            RootNodeInfo tempRoot = new(RootNodeType.Connection)
                            {
                                PasswordString = password
                            };

                            ConnectionInfo clonedTarget = exportTarget.Clone();

                            if (exportTarget is RootNodeInfo)
                            {
                                if (clonedTarget is ContainerInfo container)
                                {
                                    tempRoot.AddChildRange(container.Children.ToArray());
                                }
                                exportTarget = tempRoot;
                            }
                            else
                            {
                                tempRoot.AddChild(clonedTarget);
                                exportTarget = clonedTarget;
                            }
                        }

                        ICryptographyProvider cryptographyProvider = new CryptoProviderFactoryFromSettings().Build();
                        RootNodeInfo? rootNode = exportTarget.GetRootParent() as RootNodeInfo;
                        XmlConnectionNodeSerializer28 connectionNodeSerializer = new(
                                                                                         cryptographyProvider,
                                                                                         rootNode?.PasswordString
                                                                                                 .ConvertToSecureString() ??
                                                                                         new RootNodeInfo(RootNodeType
                                                                                                              .Connection)
                                                                                             .PasswordString
                                                                                             .ConvertToSecureString(),
                                                                                         saveFilter);
                        serializer = new XmlConnectionsSerializer(cryptographyProvider, connectionNodeSerializer)
                        {
                            UseFullEncryption = isEncrypted
                        };
                        break;
                    case SaveFormat.mRCSV:
                        serializer =
                            new CsvConnectionsSerializerMremotengFormat(saveFilter, Runtime.CredentialProviderCatalog);
                        break;
                    case SaveFormat.mRJSON:
                        serializer = new JsonConnectionsSerializer(saveFilter);
                        break;
                    case SaveFormat.RDP:
                        serializer = new RdpConnectionSerializer(saveFilter);
                        break;
                    default:
                        throw new ArgumentOutOfRangeException(nameof(saveFormat), saveFormat, null);
                }

                string serializedData = serializer.Serialize(exportTarget);
                FileDataProvider fileDataProvider = new(fileName);
                fileDataProvider.Save(serializedData);
            }
            catch (Exception ex)
            {
                Runtime.MessageCollector.AddExceptionStackTrace($"Export.SaveExportFile(\"{fileName}\") failed.", ex);
            }
            finally
            {
                Runtime.ConnectionsService.RemoteConnectionsSyncronizer?.Enable();
            }
        }
    }
}